crates/query-engine/metadata/src/metadata/native_queries.rs

```diff
@@ -46,6 +47,147 @@ pub struct ReadOnlyColumnInfo {
+            NativeQuerySqlEither::NativeQuerySqlExternal(value) => value,
+            NativeQuerySqlEither::NativeQuerySql(value) => value.into(),
+        }
+    }
+}
+
+/// A Native Query SQL after file resolution.
+/// This is the file that is expected in the metadata when translating requests.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum NativeQuerySql {
+    FromFile {
+        file: std::path::PathBuf,
+        sql: NativeQueryParts,
+    },
+    Inline {
+        sql: NativeQueryParts,
+    },
+}
+
+impl NativeQuerySql {
+    /// Extract the native query sql expression.
+    pub fn sql(self) -> NativeQueryParts {
+        match self {
+            NativeQuerySql::Inline { sql } => sql,
+            NativeQuerySql::FromFile { sql, .. } => sql,
+        }
+    }
+}
+
+// We use this type as an intermediate representation for serialization/deserialization
+// of native query sql location/expression.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+#[serde(untagged)]
+/// Native Query SQL location.
+pub enum NativeQuerySqlExternal {
+    /// Refer to an external Native Query SQL file.
+    File {
+        /// Relative path to a sql file.
+        file: std::path::PathBuf,
+    },
+    /// Inline Native Query SQL string.
+    Inline {
+        /// An inline Native Query SQL string.
+        inline: NativeQueryParts,
+    },
+    InlineUntagged(
+        /// An inline Native Query SQL string.
+        NativeQueryParts,
+    ),
+}
+
+impl NativeQuerySqlEither {
+    /// Convert an external native query sql type to NativeQuerySql,
+    /// including reading files from disk.
+    pub fn from_external(
+        &self,
+        absolute_configuration_directory: &std::path::Path,
+    ) -> Result<NativeQuerySql, String> {
+        match self {
+            // unexpected we get this, but ok.
+            NativeQuerySqlEither::NativeQuerySql(value) => Ok(value.clone()),
+            NativeQuerySqlEither::NativeQuerySqlExternal(external) => match external {
+                NativeQuerySqlExternal::File { file } => {
+                    parse_native_query_from_file(absolute_configuration_directory.join(file))
+                }
+                NativeQuerySqlExternal::Inline { inline } => Ok(NativeQuerySql::Inline {
+                    sql: inline.clone(),
+                }),
+                NativeQuerySqlExternal::InlineUntagged(inline) => Ok(NativeQuerySql::Inline {
+                    sql: inline.clone(),
+                }),
+            },
+        }
+    }
+}
+
+impl From<NativeQuerySql> for NativeQuerySqlExternal {
+    /// used for deserialization.
+    fn from(value: NativeQuerySql) -> Self {
+        match value {
+            NativeQuerySql::Inline { sql } => NativeQuerySqlExternal::Inline { inline: sql },
+            NativeQuerySql::FromFile { file, .. } => NativeQuerySqlExternal::File { file },
+        }
+    }
+}
+
+impl JsonSchema for NativeQuerySqlEither {
```

Comment:
If you want to, you can avoid this schema impl with the [`#[schemars(with="NativeQuerySqlExternal")]`](https://graham.cool/schemars/deriving/attributes/#with) attribute on `NativeQuerySqlEither` and [`#[schemars(rename="NativeQuerySql")]`](https://graham.cool/schemars/deriving/attributes/#rename) on `NativeQuerySqlExternal`.

Although it's arguably mostly a matter of style for this example.